sklearn/metrics/pairwise.py

```diff
@@ -1961,17 +1968,18 @@ def _dist_wrapper(dist_func, dist_matrix, slice_, *args, **kwargs):
 def _parallel_pairwise(X, Y, func, n_jobs, **kwds):
     """Break the pairwise matrix in n_jobs even slices
     and compute them using multithreading."""
+    xp, _ = get_namespace(X, Y)
+    dtype_float = _find_matching_floating_dtype(X, Y, xp=xp)
 
     if Y is None:
         Y = X
-    X, Y, dtype = _return_float_dtype(X, Y)
 
     if effective_n_jobs(n_jobs) == 1:
         return func(X, Y, **kwds)
 
     # enforce a threading backend to prevent data communication overhead
     fd = delayed(_dist_wrapper)
-    ret = np.empty((X.shape[0], Y.shape[0]), dtype=dtype, order="F")
+    ret = xp.empty((X.shape[0], Y.shape[0]), dtype=dtype_float, order="F")
```

Comment:
We added the `order="F"` for better cache locality (see discussion: https://github.com/scikit-learn/scikit-learn/pull/13310#discussion_r260780874). Note that we slice on `Y` later (we could keep "C" order if we sliced on `X`) because:

> The thing is in pairwise_distances_chunked we are chunking along X and then X_chunk and Y are sent to pairwise_distances that can be parallelized with n_jobs. So we cannot chunk both pairwise_distances and pairwise_distances_chunked in the same way.

What is the best way forward here? Should I do some benchmarking to see if e.g., torch array on GPU with "C" contiguous is faster than converting to numpy and using "F" ?

cc @ogrisel

Reply:
The `order=` parameter is not part of the array API spec so we cannot rely on it in general.

However, we could probably do:

```suggestion
    # Allocate and transpose such that a given thread writes its ouput to a
    # contiguous chunk in the result array.
    # Note that `xp.empty` does not offer the `order="F"` option that `np.empty`
    # offer:
    # https://data-apis.org/array-api/latest/API_specification/generated/array_api.empty.html
    # Furthermore, we assume that all array API compatible namespace allocate
    # 2D arrays with C-contiguity conventions by default which seems to be
    # the case at the time of writing.
    ret = xp.empty((Y.shape[0], X.shape[0]), dtype=dtype_float).T
```

Reply:
Let me see if this is easily amended for when we fill out `ret`

Reply:
I think array API in general is not interested in standardizing "order" (see: https://github.com/data-apis/array-api/issues/571) and avoided standardization of strides see: https://github.com/data-apis/array-api/issues/764 , so I've amended the comment to not talk specifically about `xp.empty`.

Also made some opinionated changes to the comment but happy to amend.

Reply:
@ogrisel 
I think I made the required transpose changes in: https://github.com/scikit-learn/scikit-learn/pull/29822/commits/1017ab16cd261b610de32451fe1627f66008f8ec (it's a small change but please check this is correct)
Made comment changes in https://github.com/scikit-learn/scikit-learn/pull/29822/commits/ea99f317c5db9e143ddfe2c4e4164ed256e7c4be (note that I don't think anything is required for the `_fill_diagonal` as it is only performed for square matrices)